File: app/core/parser.py
"""
PDF Parser - Converts PDF content to scene graph representation.

# DEPRECATED: Use @giga-pdf/pdf-engine via Next.js API routes instead.
#
# This module previously used PyMuPDF (AGPL) to extract text, images, shapes,
# annotations, and form fields from PDF pages and convert them to our internal
# scene graph model. All these operations are now performed by the TypeScript
# pdf-engine package (packages/pdf-engine), which is MIT-licensed and runs in
# the Next.js process.
#
# This file is retained so that existing Celery tasks (processing_tasks.py,
# export_tasks.py) and the document_service continue to import PDFParser
# without crashing. The class now returns minimal stub scene graphs and logs
# deprecation warnings instead of performing real parsing.
#
# TODO: Remove this module once all callers have been migrated to call the
#       TS engine REST API and retrieve the scene graph from there.

import base64
import io
import logging
from typing import Any, Optional

# fitz (PyMuPDF) removed — PDF parsing now done in @giga-pdf/pdf-engine (TypeScript)
import pdfplumber  # MIT-licensed, used only for basic metadata fallback

from app.core.ocr import ocr_processor
from app.models.bookmarks import BookmarkDestination, BookmarkObject, BookmarkStyle
from app.models.document import (
    DocumentMetadata,
    DocumentObject,
    DocumentPermissions,
    EmbeddedFileObject,
)
from app.models.elements import (
    AnnotationElement,
    AnnotationStyle,
    AnnotationType,
    Bounds,
    ElementType,
    FieldProperties,
    FieldStyle,
    FieldType,
    FormFieldElement,
    ImageElement,
    ImageSource,
    ImageStyle,
    LinkDestination,
    Point,
    ShapeElement,
    ShapeGeometry,
    ShapeStyle,
    ShapeType,
    TextElement,
    TextStyle,
    Transform,
)
from app.models.layers import LayerObject
from app.models.page import Dimensions, MediaBox, PageObject, PagePreview
from app.utils.coordinates import Rect
from app.utils.helpers import generate_uuid, now_utc

logger = logging.getLogger(__name__)


class PDFParser:
    """
    Parses PDF documents into scene graph representation.

    DEPRECATED: All PDF parsing is now performed by @giga-pdf/pdf-engine (TypeScript).

    This class is kept as a compatibility shim. It returns minimal stub
    DocumentObjects (correct metadata, empty page elements) so that Celery
    tasks and the document_service continue to function during migration.

    TODO: Replace all call sites with calls to the TS engine REST API.
    """

    def __init__(self, document_id: str, base_url: str = "/api/v1"):
        """
        Initialize parser.

        Args:
            document_id: Document identifier for URL generation.
            base_url: Base URL for resource endpoints.
        """
        self.document_id = document_id
        self.base_url = base_url

    def parse_document(
        self,
        doc: object,
        extract_text: bool = True,
        extract_images: bool = True,
        include_previews: bool = True,
        enable_ocr: bool = True,
        ocr_languages: str = "fra+eng",
    ) -> DocumentObject:
        """
        Parse entire document into scene graph.

        DEPRECATED: Returns a stub DocumentObject with correct page count but
        no extracted elements. Real parsing is done by @giga-pdf/pdf-engine.

        Args:
            doc: LegacyDocumentProxy or any object with .page_count.
                 If it also has ._pdf_bytes or .tobytes(), pdfplumber will be
                 used to extract basic metadata.
            extract_text, extract_images, include_previews, enable_ocr,
            ocr_languages: Kept for API compatibility — ignored.

        Returns:
            DocumentObject: Stub scene graph.
        """
        logger.warning(
            "PDFParser.parse_document() is deprecated. "
            "Use @giga-pdf/pdf-engine via Next.js API routes for PDF parsing."
        )

        page_count = getattr(doc, "page_count", 1)
        logger.info(f"Parsing (stub) document {self.document_id} with {page_count} pages")

        # Try to extract basic metadata from the doc proxy
        metadata = self._parse_metadata(doc)

        # Build stub pages — no elements (real parsing is in @giga-pdf/pdf-engine)
        pages = []
        for page_num in range(1, page_count + 1):
            page = self.parse_page(None, page_num, include_previews=include_previews)
            pages.append(page)

        return DocumentObject(
            document_id=self.document_id,
            metadata=metadata,
            pages=pages,
            outlines=[],
            layers=[],
            embedded_files=[],
        )

    def parse_page(
        self,
        page: object,
        page_number: int,
        extract_text: bool = True,
        extract_images: bool = True,
        include_previews: bool = True,
        enable_ocr: bool = True,
        ocr_languages: str = "fra+eng",
    ) -> PageObject:
        """
        Parse a single page into scene graph.

        DEPRECATED: Returns a stub PageObject with no elements.
        Real page parsing is done by @giga-pdf/pdf-engine (TypeScript).

        Args:
            page: Ignored (kept for API compatibility).
            page_number: Page number (1-indexed).
            (remaining args kept for API compatibility — ignored)

        Returns:
            PageObject: Stub page with no elements.
        """
        page_id = generate_uuid()

        # Default A4 dimensions — the TS engine provides accurate values
        dimensions = Dimensions(width=595.0, height=842.0, rotation=0)
        media_box = MediaBox(x=0.0, y=0.0, width=595.0, height=842.0)

        preview = None
        if include_previews:
            preview = PagePreview(
                thumbnail_url=f"{self.base_url}/documents/{self.document_id}/pages/{page_number}/preview?dpi=72",
                full_url=f"{self.base_url}/documents/{self.document_id}/pages/{page_number}/preview?dpi=150",
            )

        return PageObject(
            page_id=page_id,
            page_number=page_number,
            dimensions=dimensions,
            media_box=media_box,
            crop_box=None,
            elements=[],  # Elements extracted by TS engine
            preview=preview,
        )

    def _parse_metadata(self, doc: object) -> DocumentMetadata:
        """Parse document metadata from a LegacyDocumentProxy or similar."""
        permissions = DocumentPermissions(print=True, modify=True, copy=True, annotate=True)

        page_count = getattr(doc, "page_count", 1)
        is_encrypted = getattr(doc, "is_encrypted", False)
        raw_meta = {}
        try:
            raw_meta = doc.metadata or {}  # type: ignore[union-attr]
        except Exception:
            pass

        return DocumentMetadata(
            title=raw_meta.get("title"),
            author=raw_meta.get("author"),
            subject=raw_meta.get("subject"),
            keywords=raw_meta.get("keywords", "").split(",") if raw_meta.get("keywords") else [],
            creator=raw_meta.get("creator"),
            producer=raw_meta.get("producer"),
            creation_date=None,
            modification_date=None,
            page_count=page_count,
            pdf_version=raw_meta.get("format", "PDF 1.7").replace("PDF ", ""),
            is_encrypted=is_encrypted,
            permissions=permissions,
        )

    # ---------------------------------------------------------------------------
    # Private helpers retained as stubs to avoid ImportError in any remaining
    # callers. They return empty results and log a deprecation warning.
    # ---------------------------------------------------------------------------

    def _extract_text_elements(self, page: object, page_height: float) -> list:
        """DEPRECATED stub — text extraction is done by @giga-pdf/pdf-engine."""
        return []

    def _get_page_links(self, page: object) -> list:
        """DEPRECATED stub."""
        return []

    def _extract_image_elements(self, page: object, page_number: int, page_height: float) -> list:
        """DEPRECATED stub — image extraction is done by @giga-pdf/pdf-engine."""
        return []

    def _extract_annotations(self, page: object, page_height: float) -> list:
        """DEPRECATED stub — annotation extraction is done by @giga-pdf/pdf-engine."""
        return []

    def _extract_drawings(self, page: object) -> list:
        """DEPRECATED stub — drawing extraction is done by @giga-pdf/pdf-engine."""
        return []

    def _detect_corner_radius(self, items: list, width: float, height: float) -> float:
        """DEPRECATED stub."""
        return 0.0

    def _extract_form_fields(self, page: object, page_height: float) -> list:
        """DEPRECATED stub — form field extraction is done by @giga-pdf/pdf-engine."""
        return []

    def _parse_bookmarks(self, doc: object) -> list:
        """DEPRECATED stub — bookmark parsing is done by @giga-pdf/pdf-engine."""
        return []

    def _parse_layers(self, doc: object) -> list:
        """DEPRECATED stub — layer parsing is done by @giga-pdf/pdf-engine."""
        return []

    def _parse_embedded_files(self, doc: object) -> list:
        """DEPRECATED stub — embedded file parsing is done by @giga-pdf/pdf-engine."""
        return []

    @staticmethod
    def _int_to_hex_color(color_int: int) -> str:
        """Convert integer color to hex string."""
        if color_int == 0:
            return "#000000"
        r = (color_int >> 16) & 0xFF
        g = (color_int >> 8) & 0xFF
        b = color_int & 0xFF
        return f"#{r:02X}{g:02X}{b:02X}"

    @staticmethod
    def _tuple_to_hex_color(color_tuple: tuple | list) -> str:
        """Convert RGB tuple (0-1 range) to hex string."""
        if not color_tuple:
            return "#000000"
        r = int(color_tuple[0] * 255) if len(color_tuple) > 0 else 0
        g = int(color_tuple[1] * 255) if len(color_tuple) > 1 else 0
        b = int(color_tuple[2] * 255) if len(color_tuple) > 2 else 0
        return f"#{r:02X}{g:02X}{b:02X}"

    @staticmethod
    def _normalize_font_name(font_name: str) -> str:
        """Normalize font name to standard family name."""
        # Remove common suffixes
        for suffix in ["-Bold", "-Italic", "-BoldItalic", "Bold", "Italic", ",Bold", ",Italic"]:
            font_name = font_name.replace(suffix, "")

        # Map common PDF fonts to web fonts
        font_map = {
            "Helvetica": "Helvetica",
            "Arial": "Arial",
            "TimesNewRoman": "Times New Roman",
            "Times-Roman": "Times New Roman",
            "Courier": "Courier New",
            "CourierNew": "Courier New",
        }

        return font_map.get(font_name, font_name)

File: packages/pdf-engine/src/parse/parser.ts
import { randomUUID } from 'node:crypto';
import type { DocumentObject, DocumentMetadata, BookmarkObject, PageObject } from '@giga-pdf/types';
import { PDFParseError, PDFPageOutOfRangeError } from '../errors';
import { engineLogger } from '../utils/logger';
import {
  extractMetadata,
  extractLayers,
  extractEmbeddedFiles,
  extractNamedDestinations,
} from './metadata-extractor';
import { extractTextElementsByPage } from './text-extractor';
import type { TextElement } from '@giga-pdf/types';
import { extractImageElementsByPage } from './image-extractor';
import type { ImageElement } from '@giga-pdf/types';
import { extractDrawingElementsByPage } from './drawing-extractor';
import type { ShapeElement } from '@giga-pdf/types';
import { extractAnnotationElementsByPage } from './annotation-extractor';
import type { AnnotationElement } from '@giga-pdf/types';
import { extractFormFieldsByPage } from './form-extractor';
import type { FormFieldElement } from '@giga-pdf/types';
import { extractBookmarks } from './bookmark-extractor';
import { getEngine } from '../wasm';

/** A page's coordinate frame, read from the native engine (no pdfjs). */
interface PageGeometry {
  width: number;
  height: number;
  rotation: 0 | 90 | 180 | 270;
  mediaBox: { x: number; y: number; width: number; height: number };
}

/**
 * Read every page's geometry (size, rotation, raw MediaBox) from the native
 * engine in one document open. Replaces the pdfjs `getViewport` / `page.view` /
 * `page.rotate` reads the parser used to do per page.
 */
async function extractPageGeometry(
  pdfBytes: Buffer | ArrayBuffer | Uint8Array,
): Promise<{ pageCount: number; geometryByPage: Map<number, PageGeometry> }> {
  const giga = await getEngine();
  const bytes = pdfBytes instanceof Uint8Array ? pdfBytes : new Uint8Array(pdfBytes);
  let doc: ReturnType<typeof giga.open>;
  try {
    doc = giga.open(bytes);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new PDFParseError(`Failed to parse PDF: ${message}`);
  }
  const geometryByPage = new Map<number, PageGeometry>();
  let pageCount = 0;
  try {
    pageCount = doc.pageCount();
    for (let p = 1; p <= pageCount; p++) {
      const info = doc.pageInfo(p);
      const mb = info.mediaBox;
      geometryByPage.set(p, {
        width: info.width,
        height: info.height,
        rotation: (((info.rotation % 360) + 360) % 360) as 0 | 90 | 180 | 270,
        mediaBox: { x: mb[0], y: mb[1], width: mb[2], height: mb[3] },
      });
    }
  } finally {
    doc.close();
  }
  return { pageCount, geometryByPage };
}

/** Timeout in milliseconds applied to each individual extractor. */
const EXTRACTOR_TIMEOUT_MS = 5_000;

export interface ParseOptions {
  /** Extract text blocks with positions and fonts. Default: true */
  extractText?: boolean;
  /** Extract embedded images with bounds and format info. Default: true */
  extractImages?: boolean;
  /** Extract vector drawings and shapes. Default: true */
  extractDrawings?: boolean;
  /** Extract annotations (highlights, notes, links). Default: true */
  extractAnnotations?: boolean;
  /**
   * Extract AcroForm fields.
   * Alias kept for backward-compatibility: also accepted as `extractFormFields`.
   * Default: true
   */
  extractForms?: boolean;
  /** @deprecated Use `extractForms` instead. Kept for backward-compatibility. */
  extractFormFields?: boolean;
  /** Extract bookmark outlines (TOC). Default: true */
  extractBookmarks?: boolean;
  /** Generate page thumbnails. Default: false */
  includeThumbnails?: boolean;
  /** Scale factor used when generating thumbnails (0 < scale ≤ 1). Default: 0.2 */
  thumbnailScale?: number;
  /** Maximum number of pages to process. Default: 500 */
  maxPages?: number;
  /** Base URL used to build image src URLs. */
  baseUrl?: string | null;
  /** Document ID forwarded to image URLs. */
  documentId?: string;
  /** Explicit list of page numbers to parse (1-based). Overrides maxPages. */
  pages?: number[];
}

export interface ParsePageOptions {
  extractText?: boolean;
  extractImages?: boolean;
  extractDrawings?: boolean;
  extractAnnotations?: boolean;
  extractFormFields?: boolean;
  baseUrl?: string | null;
  documentId?: string;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Race a promise against a timeout.  If the timeout fires first the promise
 * rejects with a TimeoutError so the caller can apply graceful-degradation.
 */
function withTimeout<T>(promise: Promise<T>, ms: number, label: string): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`Extractor timed out after ${ms}ms: ${label}`)),
      ms,
    );
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (error: unknown) => {
        clearTimeout(timer);
        reject(error);
      },
    );
  });
}

/**
 * Run an extractor with a timeout and return an empty array on failure so that
 * a single broken extractor never blocks the whole parse pipeline.
 */
async function safeExtract<T>(
  label: string,
  fn: () => Promise<T[]>,
): Promise<T[]> {
  try {
    return await withTimeout(fn(), EXTRACTOR_TIMEOUT_MS, label);
  } catch (error) {
    engineLogger.warn(`[parser] extractor "${label}" failed — continuing without it`, {
      error: error instanceof Error ? error.message : String(error),
    });
    return [];
  }
}

/**
 * Resolve the effective boolean flag for form-field extraction.
 * Accepts both the new `extractForms` and the legacy `extractFormFields` names.
 */
function shouldExtractForms(options: ParseOptions): boolean {
  if (options.extractForms !== undefined) return options.extractForms !== false;
  if (options.extractFormFields !== undefined) return options.extractFormFields !== false;
  return true;
}

// ---------------------------------------------------------------------------
// Page-level orchestration
// ---------------------------------------------------------------------------

async function buildPageObject(
  geometry: PageGeometry,
  pageNumber: number,
  options: ParsePageOptions,
  formsByPage: Map<number, FormFieldElement[]> = new Map(),
  imagesByPage: Map<number, ImageElement[]> = new Map(),
  annotationsByPage: Map<number, AnnotationElement[]> = new Map(),
  drawingsByPage: Map<number, ShapeElement[]> = new Map(),
  textsByPage: Map<number, TextElement[]> = new Map(),
): Promise<PageObject> {
  const { width: pageWidth, height: pageHeight, rotation, mediaBox } = geometry;

  const [textElements, imageElements, drawingElements, annotationElements, formFieldElements] =
    await Promise.all([
      options.extractText !== false
        ? Promise.resolve(textsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
      options.extractImages !== false
        ? Promise.resolve(imagesByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
      options.extractDrawings !== false
        ? Promise.resolve(drawingsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
      options.extractAnnotations !== false
        ? Promise.resolve(annotationsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
      options.extractFormFields !== false
        ? Promise.resolve(formsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
    ]);

  const elements = [
    ...textElements,
    ...imageElements,
    ...drawingElements,
    ...annotationElements,
    ...formFieldElements,
  ];

  return {
    pageId: randomUUID(),
    pageNumber,
    dimensions: { width: pageWidth, height: pageHeight, rotation },
    mediaBox,
    cropBox: null,
    elements,
    preview: { thumbnailUrl: null, fullUrl: null },
  };
}

/**
 * Build a page object with per-extractor graceful degradation and timeouts.
 *
 * Extraction order (all run in parallel via Promise.all):
 *   1. text        — text blocks with positions and fonts
 *   2. images      — embedded raster images
 *   3. drawings    — vector shapes and paths
 *   4. annotations — highlights, notes, links
 *   5. forms       — AcroForm widget fields
 *
 * If any individual extractor throws or times out (5 s), its result is
 * replaced by an empty array and a warning is logged.  The other extractors
 * continue unaffected.
 */
async function buildPageObjectSafe(
  geometry: PageGeometry,
  pageNumber: number,
  options: ParseOptions,
  formsByPage: Map<number, FormFieldElement[]> = new Map(),
  imagesByPage: Map<number, ImageElement[]> = new Map(),
  annotationsByPage: Map<number, AnnotationElement[]> = new Map(),
  drawingsByPage: Map<number, ShapeElement[]> = new Map(),
  textsByPage: Map<number, TextElement[]> = new Map(),
): Promise<PageObject> {
  const { width: pageWidth, height: pageHeight, rotation, mediaBox } = geometry;

  const extractFormsEnabled = shouldExtractForms(options);

  // All five element collections are document-level (extracted once via the
  // native engine and sliced per page); forms keep their own guard upstream.
  const [textElements, imageElements, drawingElements, annotationElements, formFieldElements] =
    await Promise.all([
      options.extractText !== false
        ? Promise.resolve(textsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),

      options.extractImages !== false
        ? Promise.resolve(imagesByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),

      options.extractDrawings !== false
        ? Promise.resolve(drawingsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),

      options.extractAnnotations !== false
        ? Promise.resolve(annotationsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),

      extractFormsEnabled
        ? Promise.resolve(formsByPage.get(pageNumber) ?? [])
        : Promise.resolve([]),
    ]);

  const elements = [
    ...textElements,
    ...imageElements,
    ...drawingElements,
    ...annotationElements,
    ...formFieldElements,
  ];

  // Thumbnail generation is intentionally deferred: a thumbnail URL would
  // require a rendering step (canvas/Playwright) which is handled by the
  // preview module.  We set the URL to null and let callers request previews
  // via the dedicated /preview API when `includeThumbnails` is true.
  // This keeps the parse pipeline fast and avoids a hard dependency on the
  // browser canvas in server contexts.
  const preview =
    options.includeThumbnails === true
      ? {
          thumbnailUrl: null, // populated by preview module on demand
          fullUrl: null,
          thumbnailScale: options.thumbnailScale ?? 0.2,
        }
      : { thumbnailUrl: null, fullUrl: null };

  return {
    pageId: randomUUID(),
    pageNumber,
    dimensions: { width: pageWidth, height: pageHeight, rotation },
    mediaBox,
    cropBox: null,
    elements,
    preview,
  };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Parse a PDF document into a complete `DocumentObject`.
 *
 * Extraction pipeline (sequential steps, parallel where safe):
 *
 * Step 1 — Open PDF via pdfjs-dist.
 * Step 2 — Extract document-level data in parallel:
 *           metadata, bookmarks (outlines), layers, embedded files,
 *           named destinations.
 * Step 3 — For each page (up to `maxPages`, or the explicit `pages` list),
 *           run all enabled per-page extractors in parallel:
 *           text → images → drawings → annotations → forms.
 *           Each extractor is individually guarded: if it throws or exceeds
 *           5 s, it returns [] and logs a warning (graceful degradation).
 * Step 4 — Assemble and return the DocumentObject.
 *
 * @param pdfBytes  - PDF content as Buffer, ArrayBuffer, or Uint8Array.
 * @param options   - Fine-grained control over what to extract.
 * @returns         Fully populated DocumentObject.
 */
export async function parseDocument(
  pdfBytes: Buffer | ArrayBuffer | Uint8Array,
  options: ParseOptions = {},
): Promise<DocumentObject> {
  const documentId = options.documentId ?? randomUUID();
  const startMs = Date.now();

  engineLogger.info('[parser] parseDocument — start', { documentId });

  // pdfjs `getDocument` may transfer (and detach) the input buffer, which would
  // leave the native-engine extractors below reading a detached ArrayBuffer.
  // Snapshot a private host copy for them BEFORE pdfjs touches the original.
  const libBytes = new Uint8Array(
    pdfBytes instanceof Uint8Array ? pdfBytes : new Uint8Array(pdfBytes),
  );

  const { pageCount, geometryByPage } = await extractPageGeometry(libBytes);

  engineLogger.debug('[parser] PDF loaded', {
    documentId,
    numPages: pageCount,
  });

  // Step 2 — Document-level extractions (parallel, each with graceful fallback).
  const [metadata, bookmarks, layers, embeddedFiles, namedDestinations] = await Promise.all([
    safeExtract('metadata', async () => {
      const m = await extractMetadata(libBytes);
      // safeExtract expects T[], so we wrap the single object
      return [m];
    }).then((arr) => arr[0] ?? {
      title: null, author: null, subject: null, keywords: [], creator: null,
      producer: null, creationDate: null, modificationDate: null,
      pageCount, pdfVersion: '1.7', isEncrypted: false,
      permissions: {
        print: true, modify: true, copy: true, annotate: true,
        fillForms: true, extract: true, assemble: true, printHighQuality: true,
      },
    }),

    options.extractBookmarks !== false
      ? safeExtract('bookmarks', () => extractBookmarks(libBytes))
      : Promise.resolve([]),

    safeExtract('layers', () => extractLayers(libBytes)),
    safeExtract('embeddedFiles', () => extractEmbeddedFiles(libBytes)),

    safeExtract('namedDestinations', async () => {
      const destinations = await extractNamedDestinations(libBytes);
      // safeExtract expects T[], wrap the record as a single-element array
      return [destinations];
    }).then((arr) => arr[0] ?? {}),
  ]);

  // Step 3 — Determine which pages to process.
  const maxPages = options.maxPages ?? 500;

  let pageNumbers: number[];
  if (options.pages && options.pages.length > 0) {
    pageNumbers = options.pages;
  } else {
    const total = Math.min(pageCount, maxPages);
    pageNumbers = Array.from({ length: total }, (_, i) => i + 1);
  }

  for (const pageNum of pageNumbers) {
    if (pageNum < 1 || pageNum > pageCount) {
      throw new PDFPageOutOfRangeError(pageNum, pageCount);
    }
  }

  if (pageCount > maxPages && !options.pages) {
    engineLogger.warn('[parser] document exceeds maxPages — truncating', {
      documentId,
      totalPages: pageCount,
      maxPages,
      processing: pageNumbers.length,
    });
  }

  engineLogger.debug('[parser] extracting pages', {
    documentId,
    pages: pageNumbers.length,
  });

  // AcroForm fields and embedded images are document-level: extract each once
  // (via the native engine) and slice per page.
  const formsByPage = shouldExtractForms(options)
    ? await extractFormFieldsByPage(libBytes)
    : new Map<number, FormFieldElement[]>();
  const imagesByPage =
    options.extractImages !== false
      ? await extractImageElementsByPage(libBytes, options.baseUrl, documentId)
      : new Map<number, ImageElement[]>();
  const annotationsByPage =
    options.extractAnnotations !== false
      ? await extractAnnotationElementsByPage(libBytes)
      : new Map<number, AnnotationElement[]>();
  const drawingsByPage =
    options.extractDrawings !== false
      ? await extractDrawingElementsByPage(libBytes)
      : new Map<number, ShapeElement[]>();
  const textsByPage =
    options.extractText !== false
      ? await extractTextElementsByPage(libBytes)
      : new Map<number, TextElement[]>();

  // Step 3 — Per-page extraction (pages run in parallel; element maps are
  // sliced per page in buildPageObjectSafe).
  const pageObjects = await Promise.all(
    pageNumbers.map((pageNum) =>
      buildPageObjectSafe(
        geometryByPage.get(pageNum)!,
        pageNum,
        { ...options, documentId },
        formsByPage,
        imagesByPage,
        annotationsByPage,
        drawingsByPage,
        textsByPage,
      ),
    ),
  );

  const elapsedMs = Date.now() - startMs;
  engineLogger.info('[parser] parseDocument — complete', {
    documentId,
    pages: pageObjects.length,
    elapsedMs,
  });

  return {
    documentId,
    metadata,
    pages: pageObjects,
    outlines: bookmarks,
    namedDestinations,
    embeddedFiles,
    layers,
  };
}

export async function parsePage(
  buffer: Buffer | ArrayBuffer | Uint8Array,
  pageNumber: number,
  options: ParsePageOptions = {},
): Promise<PageObject> {
  const libBytes = new Uint8Array(
    buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer),
  );

  const { pageCount, geometryByPage } = await extractPageGeometry(libBytes);

  if (pageNumber < 1 || pageNumber > pageCount) {
    throw new PDFPageOutOfRangeError(pageNumber, pageCount);
  }

  const formsByPage =
    options.extractFormFields !== false
      ? await extractFormFieldsByPage(libBytes)
      : new Map<number, FormFieldElement[]>();
  const imagesByPage =
    options.extractImages !== false
      ? await extractImageElementsByPage(libBytes, options.baseUrl, options.documentId)
      : new Map<number, ImageElement[]>();
  const annotationsByPage =
    options.extractAnnotations !== false
      ? await extractAnnotationElementsByPage(libBytes)
      : new Map<number, AnnotationElement[]>();
  const drawingsByPage =
    options.extractDrawings !== false
      ? await extractDrawingElementsByPage(libBytes)
      : new Map<number, ShapeElement[]>();
  const textsByPage =
    options.extractText !== false
      ? await extractTextElementsByPage(libBytes)
      : new Map<number, TextElement[]>();

  return buildPageObject(
    geometryByPage.get(pageNumber)!,
    pageNumber,
    options,
    formsByPage,
    imagesByPage,
    annotationsByPage,
    drawingsByPage,
    textsByPage,
  );
}

export async function parseMetadata(
  buffer: Buffer | ArrayBuffer | Uint8Array,
): Promise<DocumentMetadata> {
  return extractMetadata(buffer);
}

export async function parseBookmarks(
  buffer: Buffer | ArrayBuffer | Uint8Array,
): Promise<BookmarkObject[]> {
  return extractBookmarks(buffer);
}
